tasks/eva_3372/src/main/groovy/eva3372/CollectSplitsAndMerges.groovy

```diff
@@ -0,0 +1,178 @@
+        def cveHashes = svesGroupedByCveHash.keySet()
+        def existingRecords = devEnv.mongoTemplate.find(query(where("_id").in(cveHashes.collect { "${assembly}#${it}".toString() })), RsIdsWithCveHash.class)
+
+        // Add to existing records
+        if (!existingRecords.isEmpty()) {
+            boolean needsUpdate = false
+            def bulkOps = devEnv.mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED, RsIdsWithCveHash.class)
+            existingRecords.each { rsIdsWithHash ->
+                def svesToProcess = svesGroupedByCveHash.get(rsIdsWithHash.cveHash)
+                def rsIdsToAdd = svesToProcess.collect { it.clusteredVariantAccession }
+                if (!rsIdsWithHash.rsIds.containsAll(rsIdsToAdd)) {
+                    rsIdsWithHash.rsIds.addAll(rsIdsToAdd)
+                    bulkOps.updateOne(query(where("_id").is(rsIdsWithHash.id)), Update.update("rsIds", rsIdsWithHash.rsIds))
+                    needsUpdate = true
+                }
+            }
+            if (needsUpdate) {
+                def bulkResult = bulkOps.execute()
+                if (bulkResult.modifiedCountAvailable && bulkResult.modifiedCount > 0) {
+                    logger.info("Updated ${bulkResult.modifiedCount} merge candidate documents (RsIdsWithCveHash)")
+                }
+            }
+        }
+
+        // Create new records
+        def processedCveHashes = existingRecords.collect { it.cveHash }
+        def cveHashesToProcess = cveHashes.findAll { !processedCveHashes.contains(it) }
+        def recordsToInsert = []
+        svesGroupedByCveHash.each { hash, svesWithHash ->
+            if (cveHashesToProcess.contains(hash)) {
+                def rsIds = svesWithHash.collect { it.clusteredVariantAccession }.toSet()
+                recordsToInsert.add(new RsIdsWithCveHash("${assembly}#${hash}", assembly, hash, rsIds))
+            }
+        }
+        if (!recordsToInsert.isEmpty()) {
+            def insertedCount = devEnv.bulkInsertIgnoreDuplicates(recordsToInsert, RsIdsWithCveHash.class)
+            logger.info("Inserted ${insertedCount} merge candidate documents (RsIdsWithCveHash)")
+        }
+    }
+
+    /**
+     * Second step to detecting splits: remove records for rsIDs without multiple CVE hashes associated
+     */
+    def detectPendingSplits = {
+        def possibleSplitCursor = new RetryableBatchingCursor<>(where("_id").regex(/^${assembly}#/), devEnv.mongoTemplate, CveHashesWithRsId.class)
+        possibleSplitCursor.each { List<CveHashesWithRsId> batch ->
+            def recordsToRemove = batch.findAll { it.cveHashes.size() < 2 }
+            def removed = devEnv.mongoTemplate.findAllAndRemove(query(where("_id").in(recordsToRemove.collect {it.id })), CveHashesWithRsId.class)
+            if (!removed.isEmpty()) {
+                logger.info("Removed ${removed.size()} split candidate documents (CveHashesWithRsId)")
+            }
+        }
```

Comment:
Wouldn't it be faster to find the on with more than one hash and copy them somewhere else then delete the whole collection ?

Reply:
Yes this is a good point, it looks like the removal took about 3 of the 8 days in the running time so this should help... I guess we'll discuss whether we want to use this script for the other assemblies, and I'll modify accordingly.

Reply:
Updated and renamed the existing collections in dev to `pendingSplits` and `pendingMerges`